libmscore/figuredbass.cpp

```diff
@@ -1770,6 +1773,256 @@ FiguredBass* Score::addFiguredBass()
+      
+void FiguredBass::realizeFiguredBass(Score *score) {
+      Q_ASSERT(score);
 
+      for (Measure* measure = score->firstMeasure(); measure; measure = measure->nextMeasure()) {
+            realizeFiguredBass(measure);
+      }
+}
+      
+void FiguredBass::realizeFiguredBass(Measure *measure) {
+      Q_ASSERT(measure);
+      const Segment::Type st = Segment::Type::ChordRest;
+      for (Segment* segment = measure->first(st); segment; segment = segment->next(st)) {
+            realizeFiguredBass(segment);
+      }
+}
+      
+void FiguredBass::realizeFiguredBass(Segment *segment) {
+      Q_ASSERT(segment);
+      int etrack = segment->score()->nstaves() * VOICES;
+
+      for (int track = 0; track < etrack; ++track) {
+            Element *element = segment->element(track);
+            if ( element == 0)
+                  continue;
+            if ( element->type() != Element::Type::CHORD)
+                  continue;
+            {
+                  vector<int> harmonys;
+                  Chord *chord = static_cast<Chord*>(element);
+                  if ( calcFiguredBass(chord, harmonys) ) { // appends entries into harmonys corresponding musical harmony pitches
+                        for (int pitch : harmonys) {
+                              NoteVal *nval = new NoteVal(pitch);
+                              chord->score()->addNote(chord,*nval);
+                        }
+                  } else {
+                        //qDebug("no figured bass on chord on track %d", chord->track());
+                  }
+            }
+      }
+}
+      
+// populate the given harmonys vector with the pitches of the harmony notes implied by the figured base on the given chord
+bool FiguredBass::calcFiguredBass(Chord *chord, vector<int> & harmonys) {
+      Q_ASSERT(harmonys.size() == 0);
+      Segment *segment = chord->segment();
+      int octave = 0; // 0, or 12 if we want the harmony 1 octave up
+      FiguredBassItem::Modifier none = FiguredBassItem::Modifier::NONE;
+      if (find_if(segment->annotations().cbegin(),
+                  segment->annotations().cend(),
+                  [chord](Element *e) {
+                        return ( e->type() == Element::Type::FIGURED_BASS
+                                && e->track() == chord->track());
+                  }) == segment->annotations().cend())
+      {
+             return false; // no figured base to annotate
+      }
+      
+      if ( chord->notes().size() == 0 ) {
+            qDebug("no notes on chord");
+            return false; // no notes in the chord, not sure if this is possible?
+      }
+      if ( chord->notes().size() != 1 ) {
+            qDebug("measure #%d, cannot render Figured Base on a multi-note chord", segment->measure()->no());
+            return false;
+      }
+      Note *note = chord->notes()[0];
+
+      auto mod7 = [] (int n) {
+            n = n%7;
+            while ( n < 0)
+                  n += 7;
+            return n;
+      };
```

Comment:
Partly for other duties, partly to think again and see if I am overlooking something, I took some time before answering. The question is:

Why using local functions at all? As they:
- make the execution flow harder to understand
- lend to creating huge, complex 'parent' functions, which could be split into smaller pieces, easier to understand and maintain, with exactly the same result
- lend to code duplication (your local function `mod7()` is likely to be needed elsewhere in a programme dealing with musical scales, but the code would need to be repeated there)
- possibly make the life of IDE's and debuggers harder

As @lasconic said, it is (mostly) a matter of taste, but I believe it is also a matter of economy: I fail to see any advantage and I feel several disadvantages in using local functions, unless they are absolutely necessary (a case, incidentally, I still have to meet). In addition, as lasconic also said, they are (almost?) never used in current MuseScore code base.

**Very minor**: why `auto`? can that function return anything else than an `int`? then, why obfuscating this piece of info under the `auto` return type?

Your code might be crystal clear and obvious to you -- as mine looks obvious to me --, but one of the tenets of open source code is that anybody else (including yourself six months from now) might be looking at your code in the future (to extend or simply maintain it) and making that anybody's life as easier as possible will help the life of the project.

I’m aware that my own code is far from obvious here and there (this is why I tend to exceed in comments) and regularly someone points these places out and starts asking embarassing questions; which is irritating, sure, but usually means that something is not adequate in those places.